### PR TITLE
TinyDNS import: Improve handling of TXT and SPF generic records

### PR DESCRIPTION
**Pull request by oyvindaabling**
The length byte(s) doesn't have to be octal escaped if
printable ascii (except colon ':' and backslash '\').

A TXT record can contain multiple strings of 1 to 127 bytes,
each with a preceding length byte (octal escaped or not).

Checklist:
- [ ] docs updated
- [ x] tests updated

### CODE REVIEW

File: server/lib/NicToolServer/Import/tinydns.pm [hunk 1]
```diff
@@ -392,6 +411,29 @@ sub unpack_domain {
     die "domain name unpack failed\n";
 }
 
+sub unpack_txt {
+    my $self = shift;
+    my $rdata = shift or die;
+
+    # Gen type 16 (TXT) rdata contains one or more strings of between 1 and
+    # 127 bytes (inclusive, after unescape), each preceeded by a length byte.
```
Comment (msimerson) [inline server/lib/NicToolServer/Import/tinydns.pm]: Where did you find this 127 byte length limit for DNS text strings? RFC 1035 limits TXT strings to 255 bytes (256 - the preceding length octet).

**Comment by oyvindaabling** [comment]
On Fri, 17 Mar 2017, Matt Simerson wrote:

> msimerson commented on this pull request.
> 
> 
> 
> > @@ -392,6 +411,29 @@ sub unpack_domain {
>      die "domain name unpack failed\n";
>  }
>  
> +sub unpack_txt {
> +    my $self = shift;
> +    my $rdata = shift or die;
> +
> +    # Gen type 16 (TXT) rdata contains one or more strings of between 1 and
> +    # 127 bytes (inclusive, after unescape), each preceeded by a length byte.
> 
> Where did you find this 127 byte length limit for DNS text strings? RFC
> 1035 limits TXT strings to 255 bytes (256 - the preceding length octet).

Good question :-)

It's a (partial) TinyDNS limitation.

The web based TinyDNS record builder we've been using doesn't emit
strings > 127, and parts of TinyDNS also doesn't use strings > 127.

Also, our DNS records, which seems to utilize (or stress :-)
the TinyDNS formats in many other aspects, doesn't
contain any TXTs with (sub) strings > 127.

... so I've simply assumed that 127 was the general
TXT limit in TinyDNS, but apparantly not (always).


Lookup the TXT records nictool1.test.deic.dk to nictool4.test.deic.dk:

* nictool1 is the #1 test string in server/t/25_import_tinydns.t,
sub test_unpack_txt (exactly - well, except for the name).

* nictool2 is the exact same TXT string, but using one
long ' (quote) TXT string, with only the colons escaped.
Notice that it's split @ 127 - TinyDNS did that.

* I hadn't actually tested it with :16: strings > 127, but now I
have, and it seems to work just fine - see the nictool4 record :-)

* nictool3 is the small double-oh-seven record in test_unpack_txt,
test #2 that stresses earlier versions of the TXT :16:
importer: octal escaped length byte followed by digits.


YMMV, depending on which patches are applied to your TinyDNS ...


unpack_txt() doesn't check for the 127 byte limit, so
it will unpack TXTs with strings > 127 bytes just fine.
I'll change the code comments to reflect this info ...

I see that the exporter emits the ' (quote) variant of the
TXT record, which doesn't have any problems with long strings.
TinyDNS splits those up automatically - with a
127 byte limit, as seen with the nictool4 TXT.

> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/msimerson/NicTool/pull/171#pullrequestreview-27665595

- -

Øyvind Aabling

DeIC / i2 | DTU, Bygning 304 | DK-2800 Lyngby
Phone: +45 35 88 82 15
Oyvind.Aabling@DeiC.dk | www.deic.dk
Oyvind.Aabling@i2.dk   | www.i2.dk

**Comment by msimerson** [comment]
> unpack_txt() doesn't check for the 127 byte limit, so it will unpack TXTs with strings > 127 bytes just fine.

I noticed. :-)

> I'll change the code comments to reflect this info

Perfect.

I've also [fixed the tinydns SPF export function](https://github.com/msimerson/NicTool/pull/173/files) to split those into 255 byte strings:

And to reinforce that the 127 bit thing is a tinydns-ism, compare these:

````
$ dig TXT apr2013._domainkey.cadillac.net. @ns2.cadillac.net. +short
"v=DKIM1\;p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA5BgmgaBHxIQSHCzFlkJ8/dCYFgppfpGxQB2mbB/nlspbZZPQPX7JrS8fglt6lVOQ/A82ErayWiABQd6GziiHbe+mA5glQSxG2o2LUtDa1AU269W1sZrgVEFkIq5sZ+T+s3KbcSjca21YOZt8NWxw5UvP1xTRHHO77JbcwUEB4rBAiZOs8eU9kuMLAuh8AQw0w17JW0+tN" "SNSphz0dY5S/5upHSdRqyOVrCJNE/Zuyzo1Ck+T1NIPt4ttd1VPkAMnjqXXjBQWP4BRObVEdmRqCxy4CRfbbiPJiNcut+iV2YezJqsVxBXwPFfsMwVb68aAHKKpdwrBfmNfv/yLdXY6RwIDAQAB"
$ dig TXT apr2013._domainkey.cadillac.net. @ns3.cadillac.net. +short
"v=DKIM1\;p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA5BgmgaBHxIQSHCzFlkJ8/dCYFgppfpGxQB2mbB/nlspbZZPQPX7JrS8fglt6lVOQ/A82ErayW" "iABQd6GziiHbe+mA5glQSxG2o2LUtDa1AU269W1sZrgVEFkIq5sZ+T+s3KbcSjca21YOZt8NWxw5UvP1xTRHHO77JbcwUEB4rBAiZOs8eU9kuMLAuh8AQw0w17JW0+t" "NSNSphz0dY5S/5upHSdRqyOVrCJNE/Zuyzo1Ck+T1NIPt4ttd1VPkAMnjqXXjBQWP4BRObVEdmRqCxy4CRfbbiPJiNcut+iV2YezJqsVxBXwPFfsMwVb68aAHKKpdwr" "BfmNfv/yLdXY6RwIDAQAB"
````

**Comment by oyvindaabling** [comment]
On Fri, 17 Mar 2017, Matt Simerson wrote:

> > unpack_txt() doesn't check for the 127 byte limit, so it will unpack TXTs with strings > 127 bytes just fine.
> 
> I noticed. :-)
> 
> > I'll change the code comments to reflect this info
> 
> Perfect.

I've updated my tinydns_import_txt branch (pull request #171).

https://github.com/oyvindaabling/NicTool/tree/tinydns_import_txt,
subject changed to
"TinyDNS import: Improve handling of TXT and SPF generic records".

New changes:

* Comments about TXT string max length changed,

* sub zr_spf changed to use unpack_txt() instead of unescape_octal(),
  so that SPF :99: rdata is handled like TXT :16: rdata,

* Fix a small typo in the sub unescape_packed_hex comments.

These two comments in sub unescape_packed_hex:

    # convert any octals to ASCII, then unpacks to hex chars
    ...
    # inserts : after each 4 char set

are a little bit confusing: "convert" vs. "unpacks" vs. "inserts"
is a mix of tenses - either there's an 's' missing
(on convert) or there are two 's'es too many :-)

The last commit in branch tinydns_import_txt removes the 's'es
to make the texts consistent with other comments in the file,
e.g. "# convert octal escapes to ascii" in sub unescape_octal.

> I've also [fixed the tinydns SPF export function](https://github.com/msimerson/NicTool/pull/173/files) to split those into 255 byte strings:
> 
> And to reinforce that the 127 bit thing is a tinydns-ism, compare these:
> 
> ````
> $ dig TXT apr2013._domainkey.cadillac.net. @ns2.cadillac.net. +short
> "v=DKIM1\;p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA5BgmgaBHxIQSHCzFlkJ8/dCYFgppfpGxQB2mbB/nlspbZZPQPX7JrS8fglt6lVOQ/A82ErayWiABQd6GziiHbe+mA5glQSxG2o2LUtDa1AU269W1sZrgVEFkIq5sZ+T+s3KbcSjca21YOZt8NWxw5UvP1xTRHHO77JbcwUEB4rBAiZOs8eU9kuMLAuh8AQw0w17JW0+tN" "SNSphz0dY5S/5upHSdRqyOVrCJNE/Zuyzo1Ck+T1NIPt4ttd1VPkAMnjqXXjBQWP4BRObVEdmRqCxy4CRfbbiPJiNcut+iV2YezJqsVxBXwPFfsMwVb68aAHKKpdwrBfmNfv/yLdXY6RwIDAQAB"
> $ dig TXT apr2013._domainkey.cadillac.net. @ns3.cadillac.net. +short
> "v=DKIM1\;p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA5BgmgaBHxIQSHCzFlkJ8/dCYFgppfpGxQB2mbB/nlspbZZPQPX7JrS8fglt6lVOQ/A82ErayW" "iABQd6GziiHbe+mA5glQSxG2o2LUtDa1AU269W1sZrgVEFkIq5sZ+T+s3KbcSjca21YOZt8NWxw5UvP1xTRHHO77JbcwUEB4rBAiZOs8eU9kuMLAuh8AQw0w17JW0+t" "NSNSphz0dY5S/5upHSdRqyOVrCJNE/Zuyzo1Ck+T1NIPt4ttd1VPkAMnjqXXjBQWP4BRObVEdmRqCxy4CRfbbiPJiNcut+iV2YezJqsVxBXwPFfsMwVb68aAHKKpdwr" "BfmNfv/yLdXY6RwIDAQAB"
> ````
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/msimerson/NicTool/pull/171#issuecomment-287459541

- -

Øyvind Aabling

DeIC / i2 | DTU, Bygning 304 | DK-2800 Lyngby
Phone: +45 35 88 82 15
Oyvind.Aabling@DeiC.dk | www.deic.dk
Oyvind.Aabling@i2.dk   | www.i2.dk

**Pull request merged**